internal/project/watch.go

```diff
@@ -60,3 +65,267 @@ func (w *watchedFiles[T]) update(ctx context.Context, newData T) (updated bool,
 	w.watcherID = watcherID
 	return true, nil
 }
+
+func createGlobMapper(host ProjectHost) func(data map[tspath.Path]string) []string {
+	rootDir := host.GetCurrentDirectory()
+	rootPath := tspath.ToPath(rootDir, "", host.FS().UseCaseSensitiveFileNames())
+	rootPathComponents := tspath.GetPathComponents(string(rootPath), "")
+	isRootWatchable := canWatchDirectoryOrFile(rootPathComponents)
+
+	return func(data map[tspath.Path]string) []string {
+		start := time.Now()
+
+		type glob struct {
+			dir          string
+			nonRecursive bool
+		}
+
+		var globSet core.Set[glob]
+
+		for path, fileName := range data {
+			w := getDirectoryToWatchFailedLookupLocation(
+				fileName,
+				path,
+				rootDir,
+				rootPath,
+				rootPathComponents,
+				isRootWatchable,
+				rootDir,
+				true,
+			)
+			if w == nil {
+				continue
+			}
+			globSet.Add(glob{dir: w.dir, nonRecursive: w.nonRecursive})
+		}
+
+		globs := make([]string, 0, globSet.Len())
+		for g := range globSet.Keys() {
+			if g.nonRecursive {
+				globs = append(globs, g.dir+"/"+fileGlobPattern)
+			} else {
+				globs = append(globs, g.dir+"/"+recursiveFileGlobPattern)
+			}
+		}
+		slices.Sort(globs)
+
+		took := time.Since(start)
+		host.Log(fmt.Sprintf("createGlobMapper took %s to create %d globs for %d failed lookups", took, len(globs), len(data)))
```

Comment:
```suggestion
		timeTaken := time.Since(start)
		host.Log(fmt.Sprintf("createGlobMapper took %s to create %d globs for %d failed lookups", timeTaken, len(globs), len(data)))
```